kdf/kdf.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2019 Thales e-Security, Inc
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+package kdf
+
+import (
+	"encoding/binary"
+	"crypto"
+	"errors"
+	"hash"
+	"io"
+	// Force registration of SHA1 and SHA2 families of cryptographic primitives to
+	// reduce the burden on KDF consuming packages.
+	_ "crypto/sha1"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+)
+
+var (
+	// ErrInvalidLengthParameter the KDF length parameter is invalid
+	ErrInvalidLengthParameter = errors.New("invalid length parameter")
+
+	// ErrInvalidSeedParameter a parameter is invalid.
+	ErrInvalidSeedParameter = errors.New("invalid input parameter")
+	)
+
+// Verify KDF completely implements the io.Reader interface.
+var _ io.Reader = &KDF{}
+
+// KDF key derivation context struct
+type KDF struct {
+	seed       []byte
+	other      []byte
+	length     int
+	iterations uint32
+	position   int
+	buffer     []byte
+	digester   hash.Hash
+}
+
+// i2osp 4-byte integer marshalling.
+func i2osp(i uint32) (osp []byte) {
+	osp = make([]byte, 4)
+	binary.BigEndian.PutUint32(osp, i)
+	return
+}
+
+// min select the minimum value of a and b.
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// Read read the next len(p) bytes from the KDF context.
+func (kdf *KDF) Read(p []byte) (n int, err error) {
+	// Read the minimum of everything requested or whatever's left.
+	toRead := min(len(p), kdf.length-kdf.position)
+	// When there's no data left fail.
+	if toRead == 0 {
+		err = io.EOF
+		return
+	}
+	// Use buffered data first to attempt to satisfy request.
+	if len(kdf.buffer) > 0 {
+		fromBuffer := min(len(kdf.buffer), toRead)
+		copy(p, kdf.buffer[:fromBuffer])
+		kdf.buffer = kdf.buffer[fromBuffer:]
+		n = fromBuffer
+	}
+	// Calculate the number of full hash outputs required to satisfy request.
+	iterations := ((toRead - n) + (kdf.digester.Size() - 1)) / kdf.digester.Size()
```

Comment:
Is this the same as `((toRead - n) / kdf.digester.Size()) + 1`? I find that easier to grok, but it may just be me.

Also, can we update `toRead` before we enter this part of the function. That will remove a lot of `toRead - n` statements that are hard to read.

Reply:
It is not the same. The code as written is equivalent to `math.Ceil(float64(toRead-n)/float64(kdf.digester.Size()))` I've broken this logic into a separate function to aid understanding and readability.